velox/core/PlanFragment.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/core/PlanFragment.h"
+#include "velox/core/QueryConfig.h"
+
+namespace facebook::velox::core {
+
+bool PlanFragment::canSpill(const QueryConfig& queryConfig) const {
+  if (not queryConfig.spillEnabled()) {
+    return false;
+  }
+
+  // Functor to check if a given node is aggregation node that can spill.
+  const auto hasSpillingHashAggrFunc = [](const core::PlanNode* node) {
```

Comment:
```
How about?

  const auto checkSpillableNodeCb = [&](const core::PlanNode* node) {
    if (node->name() == "Aggregation" && queryConfig.aggregationSpillEnabled()) {
      if (const auto* aggregationNode =
              dynamic_cast<const core::AggregationNode*>(node)) {
        if (aggregationNode->isFinal() || aggregationNode->isSingle()) {
          return true;
        }
      }
    }
    if (node->name() == "OrderBy" && queryConfig.orderBySpillEnabled())  {
     return true;
    }
    ...
    return false;
  };
```

Reply:
Sure, we can make one pass.

Reply:
@spershin @xiaoxmeng We should not write code that depends on the node names, e.g. let's remove node->name() == "Aggregation" and node->name() == "OrderBy" above.

Reply:
@mbasmanova So, how do we determine the node type then?
I don't see any node type enum.
Should we add one?

Reply:
> @spershin @xiaoxmeng We should not write code that depends on the node names, e.g. let's remove node->name() == "Aggregation" and node->name() == "OrderBy" above.
Yeah, directly check on name is hacky. I thought about add a canSpill() method in planNode but it is controlled by query config as well. What do you think we pass query config to PlanNode::canSpill(queryConfig) and internally the PlanNode will check if the corresponding spilling flag has been set in query config?

Reply:
@xiaoxmeng 

> add a canSpill() method in planNode but it is controlled by query config as well. What do you think we pass query config to PlanNode::canSpill(queryConfig) and internally the PlanNode will check if the corresponding spilling flag has been set in query config?

Sounds good to me.

Reply:
@mbasmanova, @xiaoxmeng 
Wouldn't it be a bit weird that a PlanNode would know about QueryConfig?
On the other hand we do make PLanFragment know about it...

Reply:
@spershin As you mentioned, PlanFragment already knows about QueryConfig. PlanNode is not that different. There is another reason for using PlanNode::canSpill(config). It is to extend this functionality to custom plan nodes. Current design does not allow custom plan nodes / operators to support spilling.

Reply:
@mbasmanova yeah. It might be easier for us to maintain and extend.